ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartUploadPartHandler.java

```diff
@@ -80,13 +213,88 @@ void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+              RestServiceErrorCode.InvalidArgs);
+        } else {
+          logger.debug("{} attempted an upload with ttl {} to {}", blobProperties.getServiceId(),
+              blobProperties.getTimeToLiveInSeconds(), descriptor);
+          frontendMetrics.ttlNotCompliantError.inc();
+          restResponseChannel.setHeader(RestUtils.Headers.NON_COMPLIANCE_WARNING,
+              "TTL < " + frontendConfig.maxAcceptableTtlSecsIfTtlRequired + " will be required for future uploads");
+        }
+      }
+      // inject encryption frontendMetrics if applicable
+      if (blobProperties.isEncrypted()) {
+        restRequest.getMetricsTracker()
+            .injectMetrics(frontendMetrics.putBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), true));
+      }
+      Map<String, Object> userMetadataFromRequest = new HashMap<>(restRequest.getArgs());
+      byte[] userMetadata = RestUtils.buildUserMetadata(userMetadataFromRequest);
+      frontendMetrics.blobPropsBuildForNameBlobPutTimeInMs.update(System.currentTimeMillis() - propsBuildStartTime);
+      logger.trace("Blob properties of blob being PUT - {}", blobProperties);
+      return new BlobInfo(blobProperties, userMetadata);
+    }
+
+    /**
+     * @return the {@link PutBlobOptions} to use, parsed from the request.
+     */
+    private PutBlobOptions getPutBlobOptionsFromRequest() throws RestServiceException {
+      PutBlobOptionsBuilder builder = new PutBlobOptionsBuilder().chunkUpload(false).restRequest(restRequest);
+      Long maxUploadSize = RestUtils.getLongHeader(restRequest.getArgs(), RestUtils.Headers.MAX_UPLOAD_SIZE, false);
+      if (maxUploadSize != null) {
+        builder.maxUploadSize(maxUploadSize);
       }
-    };
 
-    namedBlobPutHandler.handle(restRequest, restResponseChannel, wrappedCallback);
+      // it's the s3 part upload for the multi-part request, skip creating the composite chunk.
+      builder.skipCompositeChunk(true);
+      return builder.build();
+    }
+
+    /**
+     * Create a {@link BlobProperties} for the router upload (putBlob or stitchBlob) with a finite TTL such that
+     * orphaned blobs will not be created if the write to the named blob metadata DB fails.
+     * @param blobInfoFromRequest the {@link BlobInfo} parsed from the request.
+     * @return a {@link BlobProperties} for a TTL-ed initial router call.
+     */
+    BlobProperties getPropertiesForRouterUpload(BlobInfo blobInfoFromRequest) {
+      BlobProperties properties;
+      if (blobInfoFromRequest.getBlobProperties().getTimeToLiveInSeconds() == Utils.Infinite_Time) {
+        properties = new BlobProperties(blobInfoFromRequest.getBlobProperties());
+        // For blob with infinite time, the procedure is putBlob with a TTL, record insert to database with
+        // infinite TTL, and ttlUpdate.
+        properties.setTimeToLiveInSeconds(frontendConfig.permanentNamedBlobInitialPutTtl);
```

Comment:
Just wondering if this would also not be needed since part requests may always be temporary? But not sure though

Reply:
Leave it for now.
When we do the S3 blob life cycle management, we can revisit it.